solver/src/settlement_submission/retry.rs

```diff
@@ -0,0 +1,136 @@
+    web3::error::Error as Web3Error,
+    GasPrice,
+};
+use futures::{stream, Stream, StreamExt};
+use gas_estimation::GasPriceEstimating;
+use primitive_types::U256;
+use std::time::Duration;
+use transaction_retry::{TransactionResult, TransactionSending};
+
+// Create a never ending stream of gas prices based on checking the estimator in fixed intervals
+// and enforcing the minimum increase. Errors are ignored.
+pub fn gas_price_stream(
+    target_confirm_time: Duration,
+    gas_price_cap: f64,
+    estimator: &dyn GasPriceEstimating,
+) -> impl Stream<Item = f64> + '_ {
+    let stream = stream::unfold(true, move |first_call| async move {
+        if !first_call {
+            tokio::time::delay_for(GAS_PRICE_REFRESH_INTERVAL).await;
+        }
+        let estimate = estimator
+            .estimate_with_limits(MAX_GAS as f64, target_confirm_time)
+            .await;
+        Some((estimate, false))
+    })
+    .filter_map(|gas_price_result| async move {
+        match gas_price_result {
+            Ok(gas_price) => {
+                tracing::debug!("estimated gas price {}", gas_price);
+                Some(gas_price)
+            }
+            Err(err) => {
+                tracing::error!("gas price estimation failed: {:?}", err);
+                None
+            }
+        }
+    });
+    transaction_retry::gas_price_increase::enforce_minimum_increase_and_cap(gas_price_cap, stream)
+}
+
+fn is_transaction_error(error: &ExecutionError) -> bool {
+    // This is the error as we've seen it on openethereum nodes. The code and error messages can
+    // be found in openethereum's source code in `rpc/src/v1/helpers/errors.rs`.
+    // TODO: check how this looks on geth and infura. Not recognizing the error is not a serious
+    // problem but it will make us sometimes log an error when there actually was no problem.
+    matches!(error, ExecutionError::Web3(Web3Error::Rpc(RpcError { code, .. })) if code.code() == -32010)
+}
+
+pub struct SettleResult(pub Result<(), MethodError>);
+impl TransactionResult for SettleResult {
+    fn was_mined(&self) -> bool {
+        if let Err(err) = &self.0 {
+            !is_transaction_error(&err.inner)
+        } else {
+            true
+        }
+    }
+}
+
+pub struct SettlementSender<'a> {
+    pub contract: &'a GPv2Settlement,
+    pub nonce: U256,
+    pub settlement: EncodedSettlement,
+}
+#[async_trait::async_trait]
+impl<'a> TransactionSending for SettlementSender<'a> {
+    type Output = SettleResult;
+    async fn send(&self, gas_price: f64) -> Self::Output {
+        tracing::info!("submitting solution transaction at gas price {}", gas_price);
+        let mut method = self
+            .contract
+            .settle(
+                self.settlement.tokens.clone(),
+                self.settlement.clearing_prices.clone(),
+                self.settlement.encoded_trades.clone(),
+                self.settlement.encoded_interactions.clone(),
+                Vec::new(),
+            )
```

Comment:
nit: This code is duplicated between simulate and execute. Maybe the settlement helper struct can expose a function that returns a method builder given?